lib/rubocop/cop/lint/out_of_range_ref_in_regexp.rb

```diff
@@ -0,0 +1,71 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Lint
+      # This cops looks for out of range referencing for Regexp, as while capturing groups out of
+      # out of range reference always returns nil.
+
+      # @example
+      #   /(foo)bar/ =~ 'foobar'\
+
+      #   # bad - always returns nil
+      #   puts $2 # => nil
+
+      #   # good
+      #   puts $1 # => foo
+      #
+      class OutOfRangeRefInRegexp < Cop
+        MSG = 'Do not use out of range reference for the Regexp.'
+
+        def investigate(processed_source)
+          ast = processed_source.ast
+          valid_ref = cop_config['Count']
+          ast.each_node do |node|
+            if node.regexp_type?
+              break if contain_non_literal?(node)
+
+              tree = parse_node(node.content)
+              break if tree.nil?
+
+              valid_ref = regexp_captures(tree)
+            elsif node.nth_ref_type?
+              backref, = *node
+              add_offense(node) if backref > valid_ref
+            end
+          end
+        end
+
+        private
+
+        def contain_non_literal?(node)
+          if node.respond_to?(:type) && (node.variable? || node.send_type? || node.const_type?)
+            return true
+          end
+          return false unless node.respond_to?(:children)
+
+          node.children.any? { |child| contain_non_literal?(child) }
+        end
+
+        def parse_node(content)
+          Regexp::Parser.parse(content)
+        rescue Regexp::Scanner::ScannerError
+          nil
+        end
+
+        def regexp_captures(tree)
+          named_capture = numbered_capture = 0
+          tree.each_expression do |e|
+            named_capture += 1 if e.instance_of?(Regexp::Expression::Group::Named)
+            numbered_capture += 1 if e.instance_of?(Regexp::Expression::Group::Capture)
+          end
+          return named_capture if numbered_capture.zero?
```

Comment:
I'm not sure this is correct. You should have a test with 1 named capture and 2 numbered captures, for example...

Reply:
@marcandre, the example you have shared is mixed captures Regex, and in that case, we have to only consider named capture count as numbered captures will be ignored. so in the end, I am returning named capture count.

Just FYI, I have implemented the spec example for mix names and numbered captures.

Reply:
Oh, right, I got confused by the algorithm. I think `named_capture.non_zero? || numbered_capture` would be easier, or `named_capture > 0 ? named_capture : numbered_capture` if you prefer would be much easier to read and equivalent, right?